File: src/main/java/fr/openent/mediacentre/service/FavoriteService.java
package fr.openent.mediacentre.service;

import fr.openent.mediacentre.model.SignetResource;
import fr.openent.mediacentre.source.Signet;
import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.Optional;

public interface FavoriteService {

    /**
     * Create favorite
     * @param favorite  body favorite
     * @param handler   function handler returning da
     */
    void create(JsonObject favorite, Handler<Either<String, JsonObject>> handler);

    /**
     * Get resource with favorite
     * @param source    source parameter
     * @param userId    User identifier
     * @param handler   function handler returning da
     */
    void get(String source, String userId, Handler<Either<String, JsonArray>> handler);

    /**
     * Create favorite
     * @param favorite  body favorite
     */
    Future<JsonObject> create(JsonObject favorite);

    /**
     * Get resource with favorite
     * @param source    source parameter
     * @param userId    User identifier
     * @return {@link Future<JsonArray>} the favorites of the user
     */
    Future<JsonArray> get(String source, String userId);

    /**
     * Delete favorite
     * @param favorite  favorite identifier as parameter
     * @param source    favorite source as parameter
     * @param userId
     * @param handler   function handler returning da
     */
    void delete(String favorite, String source, String userId, Handler<Either<String, JsonObject>> handler);

    void createSQL(JsonObject favorite, String userId, Handler<Either<String, JsonObject>> defaultResponseHandler);

    void updateSQL(int favoriteId, String userId, boolean isFavorite, boolean isShare, Handler<Either<String, JsonObject>> handler);

    void getDesactivated(String signetId, JsonArray responders, Handler<Either<String, JsonArray>> handler);

    Future<Optional<SignetResource>> createSQL(JsonObject favorite, String userId);

    Future<Optional<SignetResource>> updateSQL(int signetId, String userId, boolean isFavorite, boolean isShare);

    /**
     * Delete favorite
     * @param favorite  favorite identifier as parameter
     * @param source    favorite source as parameter
     * @param userId    id of the connected user
     */
    Future<JsonObject> delete(String favorite, String source, String userId);
}

File: src/main/resources/public/ts/services/favorite.service.ts
import {ng} from 'entcore'
import http, {AxiosResponse} from 'axios';
import {IResourceBody, IResourceResponse, ResourceBody, Resource} from "../model";
import {Signet} from "../model/Signet";

export interface FavoriteService {
    create(bodyResource: IResourceBody, id:number): Promise<AxiosResponse>;
    delete(id: string, source: string): Promise<AxiosResponse>;
    get(): Promise<Array<Resource>>;
}

export const FavoriteService = ng.service('FavoriteService', (): FavoriteService => ({
    create: async (bodyResource: IResourceBody, id:number): Promise<AxiosResponse> => {
        return await http.post(`/mediacentre/favorites?id=${id}`, bodyResource);
    },

    delete: async (id: string, source: string): Promise<AxiosResponse> => {
        return await http.delete(`/mediacentre/favorites?id=${id}&source=${source}`);
    },

    get: async (): Promise<Array<Resource>> => {
        return http.get(`/mediacentre/favorites`)
            .then((response: AxiosResponse) => (response.data.data.global && response.data.data.global.length > 0) ?
                response.data.data.global.map((resource: IResourceResponse) => new Resource().build(resource)) : []);
    }
}));